pytest-embedded-idf/pytest_embedded_idf/dut.py

```diff
@@ -335,6 +347,204 @@ def parse_unity_menu_from_str(s: str) -> List[UnittestMenuCase]:
+    ):
+        if log:
+            # check format
+            check = UNITY_FIXTURE_REGEX.search(log)
+            if check:
+                regex = UNITY_FIXTURE_REGEX
+            else:
+                regex = UNITY_BASIC_REGEX
+
+            res = list(regex.finditer(log))
+        else:
+            res = []
+
+        # real parsing
+        if len(res) == 0:
+            logging.warning(f'unity test case not found, use case {case.name} instead')
+            attrs = {'name': case.name, 'result': 'FAIL', 'message': self.pexpect_proc.buffer_debug_str}
+        elif len(res) == 1:
+            attrs = {k: v for k, v in res[0].groupdict().items() if v is not None}
+        else:
+            warnings.warn('This function is for recording single unity test case only. Use the last matched one')
+            attrs = {k: v for k, v in res[-1].groupdict().items() if v is not None}
+
+        if additional_attrs:
+            attrs.update(additional_attrs)
+
+        testcase = TestCase(**attrs)
+        self.testsuite.testcases.append(testcase)
+        if testcase.result == 'FAIL':
+            self.testsuite.attrs['failures'] += 1
+        elif testcase.result == 'IGNORE':
+            self.testsuite.attrs['skipped'] += 1
+        else:
+            self.testsuite.attrs['tests'] += 1
+
+    @_record_single_unity_test_case
+    def _run_normal_case(
+        self,
+        case: UnittestMenuCase,
+        reset: bool = False,
+        timeout: float = 30,
+    ) -> None:
+        """
+        Run a specific normal case
+
+        Notes:
+            Will skip with a warning if the case type is not normal
+
+        Args:
+            case: the specific case that parsed in test menu
+            reset: whether to perform a hardware reset before running a case
+            timeout: timeout. (Default: 30 seconds)
+        """
+        if case.type != 'normal':
+            logging.warning('case %s is not a normal case', case.name)
+            return
+
+        self.expect_exact(READY_PATTERN_LIST, timeout=timeout)
+        self.write(str(case.index))
+        self.expect_exact(f'Running {case.name}...', timeout=1)
+
+    @_record_single_unity_test_case
+    def _run_multi_stage_case(
+        self,
+        case: UnittestMenuCase,
+        reset: bool = False,
+        timeout: float = 30,
+    ) -> None:
+        """
+        Run a specific multi_stage case
+
+        Notes:
+            Will skip with a warning if the case type is not normal
+
+        Args:
+            case: the specific case that parsed in test menu
+            reset: whether to perform a hardware reset before running a case
+            timeout: timeout. (Default: 30 seconds)
+        """
+        if case.type != 'multi_stage':
+            logging.warning('case %s is not a multi stage case', case.name)
+            return
+
+        _start_at = time.perf_counter()
```

Comment:
Why we still need to record the timestamp? Doesn't it have been recorded in `_record_single_unity_test_case `?

Reply:
to calculate the remain timeout.